src/xdplwf/recv.c

```diff
@@ -565,64 +642,693 @@ XdpGenericReceiveEnqueueTxNb(
+    _In_ XDP_LWF_GENERIC_RX_QUEUE *RxQueue,
+    _Inout_ NET_BUFFER_LIST *Nbl,
+    _In_ NET_BUFFER *Nb,
+    _In_ BOOLEAN CanPend,
+    _Inout_ NET_BUFFER_LIST *TxNbl
+    )
+{
+    return
+        XdpGenericRxCloneOrCopyTxNblData2(
+            RxQueue, Nbl, Nb->DataLength, Nb->DataOffset, Nb->CurrentMdl, Nb->CurrentMdlOffset,
+            CanPend, TxNbl);
+}
+
+static
+DECLSPEC_NOINLINE
+VOID
+XdpGenericRxInitializeGsoContextMdl(
+    _Inout_ NBL_RX_TX_CONTEXT *NblContext
+    )
+{
+    MmInitializeMdl(
+        &NblContext->Gso.HeaderMdl, NblContext->Gso.Headers, sizeof(NblContext->Gso.Headers));
+    MmBuildMdlForNonPagedPool(&NblContext->Gso.HeaderMdl);
+    NblContext->ValidFields.Value = 0;
+}
+
+static
+FORCEINLINE
+VOID
+XdpGenericRxInitializeGsoContext(
+    _Inout_ NBL_RX_TX_CONTEXT *NblContext,
+    _In_ UINT32 HeaderSize
+    )
+{
+    if (!NblContext->ValidFields.Gso) {
+        XdpGenericRxInitializeGsoContextMdl(NblContext);
+        NblContext->ValidFields.Gso = TRUE;
+    }
+
+    ASSERT(HeaderSize <= sizeof(NblContext->Gso.Headers));
+    NblContext->Gso.HeaderMdl.ByteCount = HeaderSize;
+}
+
+static
+_IRQL_requires_(DISPATCH_LEVEL)
+DECLSPEC_NOINLINE
+VOID
+XdpGenericRxSegmentRscToLso(
+    _In_ XDP_LWF_GENERIC_RX_QUEUE *RxQueue,
+    _Inout_ NBL_COUNTED_QUEUE *TxList,
+    _In_ NET_BUFFER_LIST *Nbl,
+    _In_opt_ const XDP_LWF_OFFLOAD_SETTING_TASK_OFFLOAD *const TaskOffload,
+    _In_ const BOOLEAN CanPend,
+    _In_ const NDIS_TCP_LARGE_SEND_OFFLOAD_NET_BUFFER_LIST_INFO LsoInfo,
+    _In_ const UINT32 TcpTotalHdrLen
+    )
+{
+    const UINT32 TcpOffset = LsoInfo.LsoV2Transmit.TcpHeaderOffset;
+    const UINT32 LsoMss = LsoInfo.LsoV2Transmit.MSS;
+    const UINT32 LsoIpVersion = LsoInfo.LsoV2Transmit.IPVersion;
+    NET_BUFFER *const Nb = Nbl->FirstNetBuffer;
+    MDL *Mdl = Nb->CurrentMdl;
+    UINT32 MdlOffset = Nb->CurrentMdlOffset;
+    ETHERNET_HEADER *const RscEthernet = RTL_PTR_ADD(Mdl->MappedSystemVa, Nb->CurrentMdlOffset);
+    TCP_HDR *const RscTcp = RTL_PTR_ADD(RscEthernet, TcpOffset);
+    const UINT8 RscThFlags = RscTcp->th_flags;
+    UINT32 SeqNum = ntohl(RscTcp->th_seq);
+    UINT32 TcpUserDataLen = Nb->DataLength - TcpTotalHdrLen;
+    UINT32 MinOffloadSize = MAXUINT32;
+    NET_BUFFER_LIST *TxNbl;
+    UINT32 SegmentsCreated = 0;
+
+    //
+    // This routine takes an arbitrary RSC frame and produces as many LSO
+    // segments as possible, followed by as many regular segments as needed.
+    // This routine also supports arbitrary hardware offload capabilities.
+    //
+
+    ASSERT(KeGetCurrentIrql() == DISPATCH_LEVEL);
+    ASSERT(!RxQueue->Flags.TxInspect);
+    ASSERT(NET_BUFFER_LIST_IS_TCP_RSC_SET(Nbl));
+    ASSERT(Nb->Next == NULL);
+    ASSERT(LsoMss > 0);
+    ASSERT(TcpTotalHdrLen <= Nb->DataLength);
+
+    //
+    // Clear any TCP flags that require special handling before they get copied
+    // into each segment.
+    //
+    RscTcp->th_flags &= ~(TH_PSH);
```

Comment:
So we only have special rules for PUSH bit?

What if the RSC frame has the FIN bit set?

Reply:
It's against the spec but I think some virtual miniports do this including netvsc when doing software RSC within netvsc.

Reply:
Yeah, I wrote this code according to the RSC and LSO specs, and also peeking at the internal software RSC library. If you've seen non-conforming NICs indicating a FIN with RSC, I can add that to be defensive here.

Reply:
We definitely hit an assert in TCP in the past where FIN bit was set on a RSC packet indicated from netvsc. So, we will need that protection.